src/fonduer/utils/udf.py

```diff
@@ -131,25 +128,35 @@ def _apply(
         for udf in self.udfs:
             udf.start()
 
+        def in_thread_func() -> None:
+            # Fill input queue with documents but # of docs in queue is capped (#435).
+            for doc in doc_loader:
+                in_queue.put(doc)  # block until a free slot is available
+                self.last_docs.add(doc.name)
+
+        Thread(target=in_thread_func).start()
+
         count_parsed = 0
+        total_count = len(doc_loader)
+
         while (
             any([udf.is_alive() for udf in self.udfs]) or not out_queue.empty()
         ) and count_parsed < total_count:
+            # Get doc from the out_queue and persist the result into postgres
             try:
-                y = out_queue.get(timeout=1)
+                y = out_queue.get()  # block until an item is available
                 self._add(y)
                 # Update progress bar whenever an item has been processed
                 count_parsed += 1
                 if self.pb is not None:
                     self.pb.update(1)
-            except Empty:
-                # This happens when any child process is alive and still processing.
-                pass
             except Exception as e:
                 # Raise an error for all the other exceptions.
                 raise (e)
 
         # Join the UDF processes
+        for _ in self.udfs:
+            in_queue.put(UDF.TASK_DONE)
```

Comment:
Just to make sure I understand, we need to put as many `UDF.TASK_DONE` into the `in_queue` here so that all of the UDFs have a signal to commit and close, correct?

Reply:
Correct. This PR changes how a child process finishes their job.
Before: they finish their job when `in_queue` becomes empty.
After: they finish their job when they receive `UDF.TASK_DONE` signal from `in_queue`.
Thus, we need to put `UDF.TASK_DONE` as many as child processes.